[cth06-Github] iP

cth06-Github opened the pull request:
# ~~Duke~~ Knight2103 Chatbot
> Light up your life with Knight2103

By Anonymous user, 5 Sep 2024

Knight2103 is a spin-off version of the original Duke chatbot. It serves to record and manage the tasks in an easy-to-manage manner, just like a Task Todo list. <br>

**Some functionalities**
1. Add Tasks to the list. 
2. Indicates if tasks is done
3. Ability to edit and delete tasks as and when needed
4. Automatically Save the tasks that you have
5. List ALL tasks to be managed where needed.

**Benefits of the application:**
- Command line interface: just type a few commands and you could use the bot!
- It's fun to use

**Functions to add**
- [X] Add functionality
- [X] Delete, edit functionality
- [X] Usable commands
- [x] Add Exceptions
- [ ] Have a GUI interface.

## Partial Reflection
> [Success is no accident. It is hard work, perseverance, learning, studying, sacrifice and most of all, love of what you are doing or learning to do](https://www.brainyquote.com/topics/perseverance-quotes). ~Pele

Indeed, this corresponds to the blockquote above. This chatbot exemplifies the intensive work and hours spent by the creator :sob: but all in all it is believed that it is worth the journey 👩‍💻 

One of the hardest part of the project is to implement the save functionality and determining how to refactor the code to OOP. Deciphering how it can change to the following (below) as indicated on the Project website is challenging.
```Java
public Duke(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            tasks = new TaskList(storage.load());
        } catch (DukeException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }

    public void run() {
        //...
    }
```
## How to use the Chatbot
`git clone` the repository into your IDE and run it to start playing. Alternatively, use the JAR file after compiling.

_Description Last Updated: 5 Sep 2024_

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

cth06-Github committed:
Rename Chatbot

cth06-Github committed:
greet & exit

cth06-Github committed:
echo success

cth06-Github committed:
add & list success

cth06-Github committed:
Task class created

cth06-Github committed:
enable isDone toggling

cth06-Github committed:
modified object string output

cth06-Github committed:
done marking success

cth06-Github committed:
add Todo class

cth06-Github committed:
added Deadline class

cth06-Github committed:
added Event class

cth06-Github committed:
case switch, level 4 complete

cth06-Github committed:
figuring out testing semi-auto

cth06-Github committed:
automate test complete

cth06-Github committed:
spacing issues

cth06-Github committed:
Exception handling done

cth06-Github committed:
delete function done

cth06-Github committed:
Add read & load files functionality

cth06-Github committed:
Improve load file functionality with Optional and exceptions

cth06-Github committed:
loadTaskList function: Use ifPresent() function

cth06-Github committed:
Update with save functionality

cth06-Github committed:
bug fix: Prevent auto-clearing of file contents

cth06-Github committed:
bug fix: Prevent tasks duplicates saved and Separate exceptions type

cth06-Github committed:
Enable append when writing files

cth06-Github committed:
Include savedTaskList.txt in .gitignore

cth06-Github committed:
Merge branch 'branch-Level-7'

cth06-Github committed:
Deadline class: Change date format from string to LocalDate

cth06-Github committed:
event class: Change time properties into LocalDateTime type

cth06-Github committed:
Add exception for catching for Event class

cth06-Github committed:
Merge branch 'branch-Level-8'

cth06-Github committed:
Create Ui class

cth06-Github committed:
Add Storage class

cth06-Github committed:
Shift code in respective classes (Parser, TaskList, Ui)

cth06-Github committed:
Refactor to OOP, include add Command class

cth06-Github committed:
Refactor code for parsing lines in file into Task

cth06-Github committed:
Update exceptions

cth06-Github committed:
Merge branch 'branch-A-MoreOOP'

cth06-Github committed:
Create and sort classess into packages

cth06-Github committed:
Merge branch 'branch-A-Packages'

cth06-Github committed:
Merge branch 'add-gradle-support'

cth06-Github committed:
Update .gitignore file to include savedTaskList.txt

cth06-Github committed:
Update gradle build

cth06-Github committed:
Add StorageTest.java

cth06-Github committed:
Add code to include exceptions for invalid integers in command

cth06-Github committed:
Update .gitignore file to include text files

cth06-Github committed:
bug fix: Change to IndexOutOfBoundsException

cth06-Github committed:
Add unit test cases

cth06-Github committed:
Update code to allow creation of Jar file

cth06-Github committed:
Add JavaDoc comments

cth06-Github committed:
Follow format style

cth06-Github committed:
Add find functionality

cth06-Github committed:
Add JavaDoc comments primarily for constructors

cth06-Github committed:
Merge branch 'branch-A-JavaDoc'

cth06-Github committed:
Resolve conflict

cth06-Github committed:
Replace wildcard imports with the specifics

cth06-Github committed:
Resolve conflicts in Parser.java

cth06-Github committed:
Resolve conflicts in Level-9

cth06-Github committed:
Change variable name taskList to tasks

cth06-Github committed:
Add more JavaDoc comments

cth06-Github committed:
Edit file

cth06-Github committed:
Merge branch 'branch-A-CodingStandard'

cth06-Github committed:
Remove redundant comments

cth06-Github committed:
Remove comment

cth06-Github committed:
Resolve conflict

Teddayz commented on src/main/java/knight2103/Knight2103.java:
```diff
@@ -0,0 +1,49 @@
+            } catch (MissingCommand e) {
+                System.out.println(e);
+            } catch (InstructionInvalid e) {
+                System.out.println("Instruction wrong format.");
```
Perhaps you could give more information about the correct instruction format so that they will know what to type in :)

cth06-Github replied:
Ah yes thanks for the suggestion!

Teddayz commented on src/main/java/knight2103/command/InstructionInvalid.java:
```diff
@@ -0,0 +1,4 @@
+package knight2103.command;
+
+public class InstructionInvalid extends Exception {
```
Perhaps can be InstructionInvalidException instead

cth06-Github replied:
Ah okay thanks!
I don't know where to write the reply but thanks for your main comment:
> Main comments your Deadline, Event Todo tasks can all be renamed to DeadlineTask, EventTask and ToDoTask so that it is clearer that they inherit from Task

It's a good suggestion.

Teddayz commented on src/main/java/knight2103/command/Parser.java:
```diff
@@ -0,0 +1,55 @@
+        String[] commandArray = fullCommand.split(" ", 2);
+        try {
+            switch (commandArray[0]) {
+            case "list":
```
I like this! It is so neat due to OOP principles being followed!

Teddayz commented on src/main/java/knight2103/files/FileContentsInvalid.java:
```diff
@@ -0,0 +1,7 @@
+package knight2103.files;
+
+public class FileContentsInvalid extends Exception {
```
Can be FileContentsInvalidException as well :)

Teddayz reviewed (commented):
Overall, I feel that it was very well done with a lot of OOP principles being followed. 
Main comments your Deadline, Event Todo tasks can all be renamed to DeadlineTask, EventTask and ToDoTask so that it is clearer that they inherit from Task 👍

Feng1231 commented on src/main/java/knight2103/Ui.java:
```diff
@@ -0,0 +1,67 @@
+import knight2103.tasks.Task;
+import knight2103.tasks.TaskList;
+
+import java.util.Scanner;
```
import order is wrong. Refer to the standards for the order

cth06-Github replied:
Oh, did the coding standards mandate any order? I thought so long as it is consistent, it's okay already.

Feng1231 commented on src/main/java/knight2103/command/AddCommand.java:
```diff
@@ -0,0 +1,57 @@
+            storage.save(taskToAdd);
+            ui.showAdd(taskToAdd, tasks);
+        } catch (DateTimeParseException e) {
+            System.out.println("knight2103.tasks.Deadline format is wrong during input. " + "Please follow " +
```
wrap before the operator

cth06-Github replied:
Nice catch, thanks!

Feng1231 commented on src/main/java/knight2103/command/ModifyCommand.java:
```diff
@@ -0,0 +1,60 @@
+            storage.save(tasks);
+        } catch (NumberFormatException e) {
+            System.out.println("Please state the task number in INTEGER. Definitely not the task name");
+        } catch (ArrayIndexOutOfBoundsException e) {
```
wrap before operator +

Feng1231 commented on src/main/java/knight2103/files/Storage.java:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return The ArrayList of task that can be stored in the bot's list of tasks.
+     */
+    public ArrayList<Task> load() {
```
hmm perhaps loadTasks() would be a clearer naming

cth06-Github replied:
Oh true, thanks for the suggestion!

Feng1231 commented on src/main/java/knight2103/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     */
+    public String printToList() {
+        String stringToPrint = "";
+        for (int i = 0; i < this.tasks.size(); i++) {
```
I believe this stringToPrint formatting could be done in the Task class.

Feng1231 commented on src/main/java/knight2103/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return Formatted List of task in the bot's application.
+     */
+    public String printToList() {
```
the return type and the method name seems to be a little off. I would expect a printToList() method to print the output in the console, instead of return a String

Feng1231 commented on src/main/java/knight2103/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return Formatted List of task to be saved in storage file.
+     */
+    public String printToFile() {
```
likewise, return and method name seems a bit off

Feng1231 commented on src/main/java/knight2103/tasks/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+     *
+     * @return Formatted list of tasks that contain the search word.
+     */
+    public String searchPrintToList(String wordSearch) {
```
likewise, return and method name seems off

Feng1231 commented on src/test/java/knight2103/files/StorageTest.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.ArrayList;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
import order is wrong. Refer to coding standard documentation

Feng1231 commented on src/test/java/knight2103/files/StorageTest.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Test
+    // ACTUALLY HOW TO WRITE ThiS TEST
+    public void load_fileInputError_someException() {
```
it seems that you handle exception for fileInputError by returning an Optional. Then I don't think this test should test for someException, as the name suggests.

cth06-Github replied:
Oh, let me take a look about that, thanks for raising it up!

Feng1231 commented on src/test/java/knight2103/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,84 @@
+
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
ordering is wrong

Feng1231 commented on src/test/java/knight2103/tasks/TaskListTest.java:
Comprehensive testing done! Wow 👍

Feng1231 reviewed (commented):
Nice! LGTM in general. The import order could be updated to match the one stated in coding standard. The naming of some methods could also be improved.

cth06-Github committed:
Set up GUI skeleton code

cth06-Github committed:
Update files to make the GUI work like the JavaFX tutorial

cth06-Github committed:
Change return type from void to String

cth06-Github committed:
Change return type from void to String for the commands

cth06-Github committed:
Allow Chatbot to show Welcome and delete some irrelevant code

cth06-Github committed:
Merge branch 'branch-Level-10'

cth06-Github committed:
Delete most commented unused code

cth06-Github committed:
ModifyCommand class: Add assertion

The if-elseif-else block works as expected only if the assumption holds
true that the code in the else block is executed only if
this.verb == CommandVerb.DELETE.

However, there may be bugs elsewhere in the programme which results
in the ModifyCommand object to be instantiated such that this.verb
accepts a CommandVerb Enum other than MARK, UNMARK or DELETE, rendering
the assumption to be invalid.

Let's add an assert statement to inform programmers when the assumption
fails to hold.

This is to ensure that bugs are caught early on and prevent an
accumulation of bugs as the programme continues to be developed.

cth06-Github committed:
Command class: Remove isExit() function

The code that drives the application in Main.java does not require to
check whether commands parsed is a kind that require exit of
application. This has simplified the codebase.

To increase readability of code, isExit() can be deleted.

Let's remove isExit() in all Command class and subclasses to improve
readability.

cth06-Github committed:
Enable FileNotFoundException message shown in GUI

When storage files to be read by the application to load the tasks are
not found, an exception is thrown and the error message is not shown in
the GUI by the chatbot, but shown in the terminal.

This is poor user feedback, undermining user experience as app users
will not know that a new file is created rather than loading a
supposedly existing file for saving or loading tasks.

Let's add the exception message into the dialog Container text parameter,
allowing the message to be shown in the GUI. To do that:
* enable the chatbot to have a welcome Message attribute which changes
  if there is an exception
* modify code in Main.java and MainWindow.java such that the UI element
  that indicates the exception message is only rendered after the
  instance of the bot is addded to the MainWindow.

This allows the Exception message to be shown in the GUI to the users.

cth06-Github committed:
Change naming of exceptions

Exception was named as FileContentsInvalid or CommandInvalid, without
the word Exception.

It may not be the most readable.

Let's change the name of exceptions.

To make the code more readable.

cth06-Github committed:
Change code to allow exceptions message to be shown in GUI (partial attempt)

cth06-Github committed:
Add Pair class

It is not possible to return two items of different types in a function
in a pair easily.

Some functions in Storage class need to return an Optional either
containing something or nothing, and an accompanying String message
which may be needed for Knight2103 as an error message.

Let's Add a Pair class which represents like a type that contains 2
different types.

So that it is possible to return 2 items of different types
simultaneously from a single function.

cth06-Github committed:
Storage class: Change return type to Pair

The bot is not be able to show all error messages to the users in the
GUI when the storage file is loaded, especially for errors that arise
due to incorrect formatting of contents in the storage file.

It is important that a comprehensive user feedback is given in the GUI,
thus code change is needed to enable that. While throwing an exception
works as well, it is not done that way because once an exception is
thrown, the entire code just stops -- i.e. the loading of file contents
will end prematurely where the next few lines in the file which are in
the correct format are not processed to create the list as comprehensive
as possible. We want to process every line in the file regardless if the
line with wrong formatting is in the middle of the correct lines.

Let's change the return type of the function such that errors in the
file contents formatting are not exactly shown as exceptions, but
saved in the bot's attribute called error message.

These error messages are created in the functions of the Storage class
which will be eventually be passed into the bot's error message
attribute, enabling the bot to show the exception message when the app
is first started.

cth06-Github committed:
Enhance input GUI functionality

When the bye command is executed, users can still interact with the
chatbot. There must be some indication that the bot is not active,
otherwise, the bye command defeats its purpose. In addition, when
users open the chatbot, the input field is automatically selected,
which obscures any placeholder text put there. Placeholder text are
important to inform users where they can write their commands in the
bot.

Let's modify the input box such that it is locked after user type in the
bye command. By doing so, user feedback about why users cannot type in
their commands should be shown to inform users. Hence, let's also have
input placeholder text to inform users as well. In summary:
* lock the input to prevent users from using further after bye command
* inform users that the input is locked and what they should do via the
  input placeholder text
* placeholder text should be enabled in the input textfield
* input textfield to be not automatically selected so that the
  placeholder text can be seen (focusTraversable attribute)

cth06-Github committed:
Remove ArrayIndexOutOfBounds Exception in catch

Catching of ArrayIndexOutOfBounds Exception in ModifyCommand class is to
catch for any errors where the command are not segmented in less than 2
parts. However, such errors should have been surfaced in the Parser.java
file and to be caught at that point in time.

Thus, there is no need to duplicate code.

Let's remove the catch statement for ArrayIndexOutOfBounds Exception in
ModifyCommand.java

This is to prevent code duplication.

cth06-Github committed:
Detect blank lines in Storage File not as errors

If the Storage textfile contains blank lines in between the list of
tasks written in the textfile, the code will detect these blank lines
as errors as they do not start with "T", "D", or "E".

This will confuse users as user feedback will say there is a need to
change the storage file contents formatting due to an error, but such
blank lines are minute deviations from the ideal formatting and may
frustrate users the limited flexibility of the formatting required in
the text file.

Let's remove such a restriction by allowing the code to check if it is a
blank line. If so, the code will automatically read the next line in the
file without further checking.

This is to allow users to be able to leave blank lines in the storage
text file, enabling flexibility in the file contents formatting,
enhancing user experience.

cth06-Github committed:
Update Exception messages shown in GUI

Messages written by the bot is not very user friendly as users do not
exactly know what went wrong. It will be better if these messages
indicate clearly for users to understand how they can input the correct
commands.

Let's edit the exception messages with more clarity.

This is to enhance user experience.

cth06-Github committed:
Refactor code via inline method

checkForMarkedFormat() is a helper function to check if the task to be
added should be marked or unmarked. If marked, the task to be added
need to have its isDone status to be updated before the task is added
into the list of tasks.

For code readability, SLAP principle is to be adhered and it seems that
such code violates SLAP principle. Since this helper function code
manages the interpretation of each line, of the same level of
abstraction as that in convertLineToTask(). Thus, this code will not be
extracted as a helper function but directly placed in
convertLineToTask().

Let's shift the code in the helper function checkForMarkedFormat()
into convertLineToTask() function.

This is to adhere to SLAP principle as much as possible.

cth06-Github committed:
Rename functions and class names for readability

Certain functions are not very clear in conveying the meaning or their
purpose. Classes as well. For example, the fact that Todo class is
inherited from Task class can be made clear by having the Todo class
being named as TodoTask. The word "print" in a function name may
suggest that the function return type is void as the statement is to be
printed only in the console. But for some functions with this "print"
word, it was meant to return a string.

Let's change the namining to convey more clearly about the
functionality and relationships of these functions and classes.

This is to improve code quality.

cth06-Github committed:
Move functions that stringify TaskList elsewhere

Functions formatToFullList(), formatToMatchedList(), formatToFileList()
are functions that make the TaskList into String format depending on
whether the list of tasks to be written is in the File Storage, or in
the GUI. And if it is in the GUI, if it is the full list of task or
selected tasks only. These functions are dependent on the context
taskList is to be shown, thus these functions may not be the best to be
in the TaskList class, and to be shifted to the Ui class and Storage
class -- the class that is relevant.

For formatToFullList() and formatToMatchedList(), both have similar code
logic which can be extracted out into a helper function, adhering to
DRY princniple.

Let's
* extract the common patterns of formatToFullList() and
  formatToMatchedList()
* shift the code to more relevant classes

This is to adhere to DRY principle and OOP design practices better.
Hopefully.

cth06-Github committed:
Remove magic numbers or literals

Some code use certain numbers or literals which its meaning may not
be the most obvious. To make the code more readable and understandable,
these numbers or literals are stored in constant variables named in
such a way that represents the meaning.

Let's replace these magic numbers or literals with constants.

This is to make the code more easy to understand.

cth06-Github committed:
Standardise import order

The order of imports are not standardised.

There is a need to standardise to enhance readability.

Let's standardise the order of imports.

This is to enhance code readability.

cth06-Github committed:
Include JavaDoc comments

JavaDoc comments are missing in some public methods or classses. This
violates coding conventions.

Let's add JavaDoc comments.

This is to inform users the functionality of the public methods.

cth06-Github committed:
Merge pull request #1 from cth06-Github/branch-A-CodeQuality

Merge branch-A-CodeQuality

cth06-Github committed:
Merge branch 'master' of https://github.com/cth06-Github/ip into branch-A-Assertions

cth06-Github committed:
Add assertions

cth06-Github committed:
Merge pull request #2 from cth06-Github/branch-A-Assertions

Add assertion round 1

cth06-Github committed:
Add yml file for Github Actions

cth06-Github committed:
Modify yml file

cth06-Github committed:
Update yml file

cth06-Github committed:
Add sorting functionality (halfway)

cth06-Github committed:
Move code on format list in Ui.java to elsewhere

formatToList(), formatToFullList(), formatToMatchedList() are private
methods in Ui class. But the Ui is to show the output of the GUI bot,
not meant to further modify the task list format. Such further
processing should be done during the execution of the command.

This is to organise the code better.

Let's
* Create FormatListCommand class which extends Command class
* Add formatToList() function in FormatListCommand class as a protected
  method
* Extend ListCommand, FindCommand class from FormatListCommand class
* Delete formatToFullList() and formatToMatchedList()

FormatListCommand class represents Commands that require formatting the
list of tasks and output the formatted list. All these commands will
need to use the formatToList(), so this function should be protected so
that commands belonging to FormatListCommand() can use this method.
formatToFullList() and formatToMatchedList() are pretty similar to
formatToList(), thus there is no need to extract it (although
technically the naming is more representative).

cth06-Github committed:
Add SortCommand to build sorting functionality

cth06-Github committed:
Move formatToList() code as a method of TaskList

formatToList(), formatToFileList() are placed in FormatListCommand and
Storage class respectively. But these methods can be a characteristic of
the taskList object representation, similar to how it is done for Task
objects. Moreover, code can be written in a more declarative style by
making use of Streams.

Let's
* Move formatToList() and formatToFileList() as methods for TaskList
  class
* Rewrite in declarative style, for loops to traverse through the list
  can be rewritten to make use of streams

This is to enhance code readability and better organisation of code.

cth06-Github committed:
Add sorting functionality

cth06-Github committed:
Storage class: Refactor code by extract method

convertLineToTask() has several aspects of repeating code, not adhering
to the DRY principle and SLAP principle.

Let's extract repeating portions as separate private methods.

This is to adhere to DRY and SLAP principles.

cth06-Github committed:
Refactor code in compare() in SortCommand class

The code did not use guard clauses and follow an ArrowHead style. This
makes code hard to read.

Let's refactor by using guard clauses.

This makes code more readable.

cth06-Github committed:
Merge branch 'branch-BCD-Extension'

cth06-Github committed:
Add command help information in welcome message

When the bot application first started, the bot will show a message at
the start.

For first time users, not having prompts as to what
commands the bot accepts right from the start is not exactly user
friendly.

Let's add some message regarding what commands users can try keying in.

This improves user experience.

cth06-Github committed:
Add condition for EventTask start time

The start date and time attribute for EventTask accepts any value, even
if it is a date and time that is after the end date and time.
Realistically, it is illogical.

To ensure that users add Event Task that makes sense, let's
* Add a condition that checks if the start date and time of the
  EventTask is before the end date and time
* Throw a DateTimeException if it is found that the start date and time
  of the EventTask is after the end date and time

This will inform users that the event task they add should be rectified.

cth06-Github committed:
Update code for better functionality

cth06-Github committed:
Polish up the code

cth06-Github committed:
Add more testing

cth06-Github committed:
Update .gitignore

cth06-Github committed:
Update README.md

cth06-Github committed:
Update README.md

cth06-Github committed:
Delete testTaskList3.txt

cth06-Github committed:
Delete 13savedTaskList.txt

cth06-Github committed:
Rectify image in README.md

cth06-Github committed:
Merge branch 'master' of https://github.com/cth06-Github/ip

cth06-Github committed:
Rectify image in README.md

cth06-Github committed:
Add code in Task.java to handle blank description

Sub-classes of Task class accepts task description that is empty or just
only whitespaces. Realistically, this means the task description is
empty and the tasks should not be created in the first place.

Let's Add a condition to ensure that task created has a non-empty task
description, i.e. task description should contain non-whitespace
characters.

This is to improve user experience and functionality.

cth06-Github committed:
Update README.md

cth06-Github committed:
Push text files used for testing

cth06-Github committed:
Clean the code

cth06-Github committed:
Update README.md

cth06-Github committed:
Add link to updated User Guide in README
